File: qprotect/identity.py
"""Validation and canonical serialization for sensitive identity data."""
from __future__ import annotations

from dataclasses import asdict, dataclass, field
from datetime import datetime
import ipaddress
import json
import re
from typing import Any, Mapping
import uuid as uuidlib

from .exceptions import IdentityValidationError

_MAC_ALIASES = ("mac", "mac_address", "hardware_mac")
_BSSID_ALIASES = ("bssid", "wifi_bssid")
_SERIAL_ALIASES = ("serial", "hardware_serial", "serial_number")
_IP_ALIASES = ("ip", "ip_address")
_UUID_ALIASES = ("uuid",)
_GUID_ALIASES = ("guid",)
_FINGERPRINT_ALIASES = ("browser_fingerprint", "browser_fingerprint_hash")
_GPS_ALIASES = ("gps", "location")


def _first(data: Mapping[str, Any], names: tuple[str, ...]) -> Any:
    for name in names:
        if name in data:
            return data[name]
    return None


def _canonical(data: Any) -> bytes:
    return json.dumps(
        data,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,
    ).encode("utf-8")


def _valid_mac(value: Any) -> str:
    if not isinstance(value, str):
        raise IdentityValidationError("MAC/BSSID must be a string")
    compact = re.sub(r"[:.\- ]", "", value.strip())
    if not re.fullmatch(r"[0-9A-Fa-f]{12}", compact):
        raise IdentityValidationError("MAC/BSSID is not a 48-bit EUI value")
    groups = [compact[index:index + 2].lower() for index in range(0, 12, 2)]
    return ":".join(groups)


def _valid_ip(value: Any) -> str:
    if not isinstance(value, str):
        raise IdentityValidationError("IP address must be a string")
    try:
        return str(ipaddress.ip_address(value.strip()))
    except ValueError as exc:
        raise IdentityValidationError("invalid IP address") from exc


def _valid_serial(value: Any) -> str:
    if not isinstance(value, str):
        raise IdentityValidationError("hardware serial must be a string")
    value = value.strip()
    if not value or len(value.encode("utf-8")) > 512:
        raise IdentityValidationError("hardware serial must be 1..512 bytes")
    if any(ord(char) < 32 or char == "\x7f" for char in value):
        raise IdentityValidationError("hardware serial contains control characters")
    return value


def _valid_uuid(value: Any) -> str:
    if not isinstance(value, str):
        raise IdentityValidationError("UUID/GUID must be a string")
    try:
        return str(uuidlib.UUID(value.strip()))
    except ValueError as exc:
        raise IdentityValidationError("invalid UUID/GUID") from exc


def _valid_fingerprint(value: Any) -> Any:
    if value is None:
        return None
    if not isinstance(value, (str, int, float, bool, list, dict)):
        raise IdentityValidationError("browser fingerprint must be JSON data")
    if isinstance(value, dict) and not all(isinstance(key, str) for key in value):
        raise IdentityValidationError("browser fingerprint object keys must be strings")
    if isinstance(value, list) and any(isinstance(item, dict) and not all(isinstance(key, str) for key in item) for item in value):
        raise IdentityValidationError("browser fingerprint object keys must be strings")
    try:
        encoded = _canonical(value)
    except Exception as exc:
        raise IdentityValidationError("browser fingerprint is not JSON-serializable") from exc
    if len(encoded) > 64 * 1024:
        raise IdentityValidationError("browser fingerprint exceeds 64 KiB")
    if isinstance(value, dict):
        return dict(sorted((str(key), item) for key, item in value.items()))
    return value


def _valid_gps(value: Any) -> dict[str, Any] | None:
    if value is None:
        return None
    if not isinstance(value, Mapping):
        raise IdentityValidationError("GPS data must be an object")
    if not all(isinstance(key, str) for key in value):
        raise IdentityValidationError("GPS data keys must be strings")
    result: dict[str, Any] = {}
    for name in ("latitude", "longitude"):
        if name not in value:
            raise IdentityValidationError(f"GPS data missing {name}")
        try:
            if isinstance(value[name], bool):
                raise TypeError
            result[name] = float(value[name])
        except (TypeError, ValueError) as exc:
            raise IdentityValidationError(f"GPS {name} is not numeric") from exc
    if not -90.0 <= result["latitude"] <= 90.0:
        raise IdentityValidationError("GPS latitude is outside [-90, 90]")
    if not -180.0 <= result["longitude"] <= 180.0:
        raise IdentityValidationError("GPS longitude is outside [-180, 180]")
    for name, minimum, maximum in (
        ("altitude_m", -10000.0, 100000.0),
        ("accuracy_m", 0.0, 1000000.0),
    ):
        if name in value and value[name] is not None:
            try:
                if isinstance(value[name], bool):
                    raise TypeError
                number = float(value[name])
            except (TypeError, ValueError) as exc:
                raise IdentityValidationError(f"GPS {name} is not numeric") from exc
            if not minimum <= number <= maximum:
                raise IdentityValidationError(f"GPS {name} is outside the accepted range")
            result[name] = number
    if "timestamp" in value and value["timestamp"] is not None:
        if not isinstance(value["timestamp"], str):
            raise IdentityValidationError("GPS timestamp must be a string")
        try:
            parsed_timestamp = datetime.fromisoformat(value["timestamp"].replace("Z", "+00:00"))
        except ValueError as exc:
            raise IdentityValidationError("GPS timestamp must be RFC 3339") from exc
        if parsed_timestamp.tzinfo is None:
            raise IdentityValidationError("GPS timestamp must include a timezone")
        result["timestamp"] = value["timestamp"]
    encoded = _canonical(result)
    if len(encoded) > 4096:
        raise IdentityValidationError("GPS data exceeds 4 KiB")
    return result


@dataclass(frozen=True)
class IdentityRecord:
    ip_address: str | None = None
    mac_address: str | None = None
    hardware_serial: str | None = None
    uuid: str | None = None
    guid: str | None = None
    browser_fingerprint: Any = None
    wifi_bssid: str | None = None
    gps: dict[str, Any] | None = None
    additional: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    def to_payload(self) -> bytes:
        return _canonical(self.to_dict())

    @classmethod
    def from_mapping(cls, data: Mapping[str, Any]) -> IdentityRecord:
        if not isinstance(data, Mapping):
            raise IdentityValidationError("identity input must be a JSON object")
        normalized_keys: dict[Any, Any] = {}
        for key, value in data.items():
            normalized = key.strip() if isinstance(key, str) else key
            if normalized in normalized_keys:
                raise IdentityValidationError("identity input contains duplicate normalized keys")
            normalized_keys[normalized] = value
        ip_value = _first(normalized_keys, _IP_ALIASES)
        mac_value = _first(normalized_keys, _MAC_ALIASES)
        bssid_value = _first(normalized_keys, _BSSID_ALIASES)
        serial_value = _first(normalized_keys, _SERIAL_ALIASES)
        uuid_value = _first(normalized_keys, _UUID_ALIASES)
        guid_value = _first(normalized_keys, _GUID_ALIASES)
        fingerprint_value = _first(normalized_keys, _FINGERPRINT_ALIASES)
        gps_value = _first(normalized_keys, _GPS_ALIASES)

        known = set(_IP_ALIASES + _MAC_ALIASES + _BSSID_ALIASES + _SERIAL_ALIASES + _UUID_ALIASES + _GUID_ALIASES + _FINGERPRINT_ALIASES + _GPS_ALIASES)
        additional: dict[str, Any] = {}
        for key, value in normalized_keys.items():
            if key in known:
                continue
            if key == "additional":
                if not isinstance(value, Mapping):
                    raise IdentityValidationError("additional identity data must be an object")
                for subkey, subvalue in value.items():
                    if not isinstance(subkey, str):
                        raise IdentityValidationError("additional identity data keys must be strings")
                    additional[subkey] = subvalue
            elif isinstance(key, str):
                additional[key] = value
            else:
                raise IdentityValidationError("additional identity data keys must be strings")
        try:
            encoded = _canonical(additional)
        except Exception as exc:
            raise IdentityValidationError("additional identity data is not JSON-serializable") from exc
        if len(encoded) > 64 * 1024:
            raise IdentityValidationError("additional identity data exceeds 64 KiB")

        return cls(
            ip_address=_valid_ip(ip_value) if ip_value is not None else None,
            mac_address=_valid_mac(mac_value) if mac_value is not None else None,
            hardware_serial=_valid_serial(serial_value) if serial_value is not None else None,
            uuid=_valid_uuid(uuid_value) if uuid_value is not None else None,
            guid=_valid_uuid(guid_value) if guid_value is not None else None,
            browser_fingerprint=_valid_fingerprint(fingerprint_value),
            wifi_bssid=_valid_mac(bssid_value) if bssid_value is not None else None,
            gps=_valid_gps(gps_value),
            additional=additional,
        )

    @classmethod
    def from_json(cls, value: str | bytes) -> IdentityRecord:
        def reject_duplicates(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
            result: dict[str, Any] = {}
            for key, item in pairs:
                if key in result:
                    raise IdentityValidationError("identity input contains duplicate JSON keys")
                result[key] = item
            return result

        def reject_constant(_: str) -> None:
            raise IdentityValidationError("identity input contains a non-finite number")

        try:
            parsed = json.loads(
                value,
                object_pairs_hook=reject_duplicates,
                parse_constant=reject_constant,
            )
        except IdentityValidationError:
            raise
        except Exception as exc:
            raise IdentityValidationError("identity input is not valid JSON") from exc
        return cls.from_mapping(parsed)


def validate_identity(data: Mapping[str, Any]) -> IdentityRecord:
    """Validate sensitive identity data without echoing it in errors."""
    return IdentityRecord.from_mapping(data)

File: cpp/src/identity.cpp
#include "identity.hpp"

#include <arpa/inet.h>

#include <algorithm>
#include <array>
#include <cctype>
#include <chrono>
#include <cmath>
#include <regex>
#include <set>
#include <tuple>

namespace qprotect::cpp::detail {
namespace {

using json::Object;
using json::Value;

std::string trim_ascii(const std::string& value) {
    std::size_t begin = 0;
    std::size_t end = value.size();
    while (begin < end && std::isspace(static_cast<unsigned char>(value[begin]))) {
        ++begin;
    }
    while (end > begin && std::isspace(static_cast<unsigned char>(value[end - 1]))) {
        --end;
    }
    return value.substr(begin, end - begin);
}

const Value* first(const Object& object, std::initializer_list<const char*> names) {
    for (const char* name : names) {
        const auto found = object.find(name);
        if (found != object.end()) {
            return &found->second;
        }
    }
    return nullptr;
}

const std::string& require_string(const Value* value, const char* field) {
    if (value == nullptr || !value->is_string()) {
        throw EnvelopeError(std::string(field) + " must be a string");
    }
    return value->as_string();
}

std::string normalize_mac(const Value* value, const char* field) {
    if (value == nullptr || value->is_null()) {
        return {};
    }
    std::string compact;
    for (const unsigned char raw : trim_ascii(require_string(value, field))) {
        if (raw == ':' || raw == '.' || raw == '-' || raw == ' ') {
            continue;
        }
        if (!std::isxdigit(raw)) {
            throw EnvelopeError(std::string(field) + " is not a 48-bit EUI value");
        }
        compact.push_back(static_cast<char>(std::tolower(raw)));
    }
    if (compact.size() != 12) {
        throw EnvelopeError(std::string(field) + " is not a 48-bit EUI value");
    }
    std::string normalized;
    for (std::size_t index = 0; index < compact.size(); index += 2) {
        if (!normalized.empty()) {
            normalized.push_back(':');
        }
        normalized.append(compact, index, 2);
    }
    return normalized;
}

std::string normalize_ip(const Value* value) {
    if (value == nullptr || value->is_null()) {
        return {};
    }
    const std::string input = trim_ascii(require_string(value, "IP address"));
    std::array<unsigned char, 16> address{};
    char rendered[INET6_ADDRSTRLEN]{};
    if (inet_pton(AF_INET, input.c_str(), address.data()) == 1) {
        if (inet_ntop(AF_INET, address.data(), rendered, sizeof(rendered)) == nullptr) {
            throw EnvelopeError("unable to normalize IP address");
        }
        return rendered;
    }
    if (inet_pton(AF_INET6, input.c_str(), address.data()) == 1) {
        if (inet_ntop(AF_INET6, address.data(), rendered, sizeof(rendered)) == nullptr) {
            throw EnvelopeError("unable to normalize IP address");
        }
        std::string normalized(rendered);
        std::transform(normalized.begin(), normalized.end(), normalized.begin(), [](unsigned char c) {
            return static_cast<char>(std::tolower(c));
        });
        return normalized;
    }
    throw EnvelopeError("invalid IP address");
}

std::string normalize_uuid(const Value* value, const char* field) {
    if (value == nullptr || value->is_null()) {
        return {};
    }
    std::string input = trim_ascii(require_string(value, field));
    std::transform(input.begin(), input.end(), input.begin(), [](unsigned char c) {
        return static_cast<char>(std::tolower(c));
    });
    if (input.starts_with("urn:uuid:")) {
        input.erase(0, 9);
    }
    if (input.size() == 38 && input.front() == '{' && input.back() == '}') {
        input = input.substr(1, 36);
    }
    std::string hex;
    for (const char c : input) {
        if (c == '-') {
            continue;
        }
        if (!std::isxdigit(static_cast<unsigned char>(c))) {
            throw EnvelopeError(std::string("invalid ") + field);
        }
        hex.push_back(c);
    }
    if (hex.size() != 32) {
        throw EnvelopeError(std::string("invalid ") + field);
    }
    return hex.substr(0, 8) + "-" + hex.substr(8, 4) + "-" + hex.substr(12, 4) + "-" +
        hex.substr(16, 4) + "-" + hex.substr(20, 12);
}

std::string normalize_serial(const Value* value) {
    if (value == nullptr || value->is_null()) {
        return {};
    }
    const std::string serial = trim_ascii(require_string(value, "hardware serial"));
    if (serial.empty() || serial.size() > 512) {
        throw EnvelopeError("hardware serial must be 1..512 bytes");
    }
    for (const unsigned char c : serial) {
        if (c < 0x20 || c == 0x7f) {
            throw EnvelopeError("hardware serial contains control characters");
        }
    }
    return serial;
}

void validate_timestamp(const std::string& text) {
    static const std::regex pattern(
        R"(^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(\.[0-9]+)?(Z|[+-][0-9]{2}:[0-9]{2})$)");
    std::string whole_seconds = text;
    const std::size_t fraction = whole_seconds.find('.', 19);
    if (fraction != std::string::npos) {
        const std::size_t zone = whole_seconds.find_first_of("Z+-", fraction + 1);
        if (zone == std::string::npos || zone == fraction + 1 ||
            !std::all_of(whole_seconds.begin() + static_cast<std::ptrdiff_t>(fraction + 1),
                         whole_seconds.begin() + static_cast<std::ptrdiff_t>(zone),
                         [](unsigned char c) { return std::isdigit(c); })) {
            throw EnvelopeError("GPS timestamp has an invalid fractional second");
        }
        whole_seconds.erase(fraction, zone - fraction);
    }
    if (!std::regex_match(whole_seconds, pattern)) {
        throw EnvelopeError("GPS timestamp must be RFC 3339 with a timezone");
    }
    const auto number = [&text](std::size_t at, std::size_t length) {
        return std::stoi(text.substr(at, length));
    };
    const std::chrono::year_month_day date{
        std::chrono::year(number(0, 4)),
        std::chrono::month(static_cast<unsigned>(number(5, 2))),
        std::chrono::day(static_cast<unsigned>(number(8, 2))),
    };
    if (!date.ok() || number(11, 2) > 23 || number(14, 2) > 59 || number(17, 2) > 59) {
        throw EnvelopeError("GPS timestamp is not a valid date and time");
    }
    const std::size_t zone = text.find_first_of("Z+-", 19);
    if (zone != std::string::npos && text[zone] != 'Z' &&
        (number(zone + 1, 2) > 23 || number(zone + 4, 2) > 59)) {
        throw EnvelopeError("GPS timestamp has an invalid timezone offset");
    }
}

Value normalize_gps(const Value* value) {
    if (value == nullptr || value->is_null()) {
        return Value(nullptr);
    }
    const Object& input = value->expect_object("GPS data");
    Object output;
    for (const char* field : {"latitude", "longitude"}) {
        const auto found = input.find(field);
        if (found == input.end() || !found->second.is_number()) {
            throw EnvelopeError(std::string("GPS data missing or invalid ") + field);
        }
        const double number = found->second.as_number();
        const double minimum = std::string(field) == "latitude" ? -90.0 : -180.0;
        const double maximum = -minimum;
        if (!std::isfinite(number) || number < minimum || number > maximum) {
            throw EnvelopeError(std::string("GPS ") + field + " is outside its accepted range");
        }
        output.emplace(field, Value(number));
    }
    for (const auto& [field, minimum, maximum] : {
             std::tuple<const char*, double, double>{"altitude_m", -10000.0, 100000.0},
             {"accuracy_m", 0.0, 1000000.0}}) {
        const auto found = input.find(field);
        if (found == input.end() || found->second.is_null()) {
            continue;
        }
        if (!found->second.is_number()) {
            throw EnvelopeError(std::string("GPS ") + field + " is not numeric");
        }
        const double number = found->second.as_number();
        if (!std::isfinite(number) || number < minimum || number > maximum) {
            throw EnvelopeError(std::string("GPS ") + field + " is outside its accepted range");
        }
        output.emplace(field, Value(number));
    }
    const auto timestamp = input.find("timestamp");
    if (timestamp != input.end() && !timestamp->second.is_null()) {
        const std::string& text = require_string(&timestamp->second, "GPS timestamp");
        validate_timestamp(text);
        output.emplace("timestamp", Value(text));
    }
    if (Value(output).canonical().size() > 4096) {
        throw EnvelopeError("GPS data exceeds 4 KiB");
    }
    return Value(std::move(output));
}

} // namespace

json::Value normalize_identity(const json::Value& input_value) {
    const Object& input = input_value.expect_object("identity input");
    Object normalized;
    for (const auto& [key, value] : input) {
        const std::string name = trim_ascii(key);
        if (!normalized.emplace(name, value).second) {
            throw EnvelopeError("identity input contains duplicate normalized keys");
        }
    }

    const std::set<std::string> known{
        "ip", "ip_address", "mac", "mac_address", "hardware_mac", "bssid", "wifi_bssid",
        "serial", "hardware_serial", "serial_number", "uuid", "guid", "browser_fingerprint",
        "browser_fingerprint_hash", "gps", "location", "additional",
    };
    Object additional;
    const auto supplied_additional = normalized.find("additional");
    if (supplied_additional != normalized.end()) {
        const Object& values = supplied_additional->second.expect_object("additional identity data");
        additional = values;
    }
    for (const auto& [key, value] : normalized) {
        if (known.contains(key)) {
            continue;
        }
        additional.insert_or_assign(key, value);
    }
    if (Value(additional).canonical().size() > 64 * 1024) {
        throw EnvelopeError("additional identity data exceeds 64 KiB");
    }

    Value fingerprint(nullptr);
    const Value* supplied_fingerprint = first(normalized, {"browser_fingerprint", "browser_fingerprint_hash"});
    if (supplied_fingerprint != nullptr) {
        fingerprint = *supplied_fingerprint;
        if (fingerprint.canonical().size() > 64 * 1024) {
            throw EnvelopeError("browser fingerprint exceeds 64 KiB");
        }
    }

    Object result;
    const std::string ip = normalize_ip(first(normalized, {"ip", "ip_address"}));
    const std::string mac = normalize_mac(first(normalized, {"mac", "mac_address", "hardware_mac"}), "MAC address");
    const std::string bssid = normalize_mac(first(normalized, {"bssid", "wifi_bssid"}), "Wi-Fi BSSID");
    const std::string serial = normalize_serial(first(normalized, {"serial", "hardware_serial", "serial_number"}));
    const std::string uuid = normalize_uuid(first(normalized, {"uuid"}), "UUID");
    const std::string guid = normalize_uuid(first(normalized, {"guid"}), "GUID");
    result.emplace("additional", Value(std::move(additional)));
    result.emplace("browser_fingerprint", std::move(fingerprint));
    result.emplace("gps", normalize_gps(first(normalized, {"gps", "location"})));
    result.emplace("guid", guid.empty() ? Value(nullptr) : Value(guid));
    result.emplace("hardware_serial", serial.empty() ? Value(nullptr) : Value(serial));
    result.emplace("ip_address", ip.empty() ? Value(nullptr) : Value(ip));
    result.emplace("mac_address", mac.empty() ? Value(nullptr) : Value(mac));
    result.emplace("uuid", uuid.empty() ? Value(nullptr) : Value(uuid));
    result.emplace("wifi_bssid", bssid.empty() ? Value(nullptr) : Value(bssid));
    return Value(std::move(result));
}

} // namespace qprotect::cpp::detail
